### PR TITLE
feat(sdk-auth): add module for performing auth flows against CTP OAuth API

### LINKED ISSUE #718: Some limitations of the sdk-auth-middleware when used in SPA+Apollo
### Description

As you may know, in SUNRISE (as Single-Page Application) we are using Apollo to communicate with the commercetools' GraphQL endpoint, so we don't use the Node.js SDK. But to avoid duplicating all the authentication logic, we are using the sdk-middleware-auth to fetch/refresh the different tokens, which we then connect to Apollo as a middleware (https://github.com/commercetools/sunrise-spa/blob/3dd595841630907f2d5efb63b355b2b8d763ffa1/src/apollo.js#L23-L29). It works well, even though it might be a bit unnatural.

Of course it's the SDK's middleware that manages the token and we don't have any access to it. That was not an issue with the Credentials Flow, but with the Password Flow that becomes really problematic because we have no way to keep the refresh token in the storage, to prevent the user from being logged out when s/he reloads the app (you can test it here: https://sunrise.netlify.com/).

So in short, the problems are:
1) Refresh tokens are lost on reloading the app, so it's not usable. Allowing to use a custom storage would be a solution.
2) It would be great to have authentication support (fetch/refresh token) designed for being used without the SDK, as sometimes other clients, such as Apollo, offer so much more specialized support that our clients are not meant to beat. Either extracting the logic into a non-middleware type of library or exporting some of the functions would be a big improvement.

Could you help us? Specially point num. 1 is high priority IMO as I don't think it's usable for anybody at the moment. Num. 2 is a really nice improvement that would allow commercetools to actively support using Apollo for GraphQL-based apps (and other clients as well).

### Expected Behavior

1) When reloading a single-page application using the sdk-auth-middleware, you should stay logged in.
2) When implementing an application where I don't want to use the SDK (e.g. I want to use Apollo), I still want commercetools to provide a tool that takes care of the authentication.

### Current Behavior

1) When reloading, it logs out, as the refresh token is lost and I have no way to get it.
2) I need to fake a request and capture the token with a Promise, which is not pretty (but it works).

### Context
I'm trying to build a single-page application using Apollo to directly connect to our GraphQL endpoint. As it's a SPA, we need to use the /me endpoints to access customer/cart information.

### Possible Solution
1) Allow to access the internal store or define where the tokens should be stored.
2) Extract the functionality into another library that is not acting as middleware, and call it from the middleware.

### PR DESCRIPTION
**Pull request by junajan**
#### Summary
This PR adds a new module which can be used for retrieving an access token from CTP OAuth API.

Resolves #718 

#### Todo

- Tests
  - [x] Unit
  - [x] Integration
  - [ ] Acceptance
- [x] Documentation

### CODE REVIEW

File: integration-tests/sdk/auth.it.js [hunk 1]
```diff
@@ -1,241 +1,265 @@
...
-import {
-  createAuthMiddlewareForPasswordFlow,
-  createAuthMiddlewareForRefreshTokenFlow,
-  createAuthMiddlewareForAnonymousSessionFlow,
-} from '@commercetools/sdk-middleware-auth'
 import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
-import fetch from 'node-fetch'
+import { createAuthMiddlewareWithExistingToken } from '@commercetools/sdk-middleware-auth'
+import { createClient } from '@commercetools/sdk-client'
+import SdkAuth from '../../packages/sdk-auth'
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth.it.js]: Hi @daern91 .. what is the benefit of using released/linked packages in integration tests?

File: integration-tests/sdk/auth-middleware.it.js [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  })
+
+  describe('Password Session Flow', () => {
+    const httpMiddleware = createHttpMiddleware({
+      host: 'https://api.sphere.io',
+      fetch,
+    })
+
+    it('should log customer and fetch customer profile', () => {
+      const userConfig = {
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth-middleware.it.js]: Hi @daern91 .. I renamed old `auth.it.js` which contains tests for `auth-middleware` to `auth-middleware.it.js` so this code here is an old code from original file.

File: integration-tests/sdk/auth-middleware.it.js [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+          ...{
+            credentials: {
+              clientId: apiConfig.credentials.clientId,
+              clientSecret: apiConfig.credentials.clientSecret,
+              anonymousId,
+            },
+          },
+          fetch,
+        }
+        const client = createClient({
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth-middleware.it.js]: As in the previous comment .. I renamed auth.it.js to auth-middleware.it.js so these are old tests

File: integration-tests/sdk/auth.it.js [hunk 4]
```diff
@@ -1,241 +1,265 @@
...
-              clientSecret: apiConfig.credentials.clientSecret,
-              anonymousId,
+    it('should introspect an invalid token', async () => {
+      const introspection = await authClient.introspectToken('invalid_token')
+      expect(introspection).toHaveProperty('active', false)
+    })
+  })
+
+  describe('Error cases', () => {
+    it('should send an invalid customer credentials', async () => {
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth.it.js]: Updated here 3a670fe

File: integration-tests/sdk/auth.it.js [hunk 5]
```diff
@@ -1,241 +1,265 @@
...
 
-  describe('Refresh Token Flow', () => {
-    it(
-      'uses the refresh token',
-      () => {
-        const httpMiddleware = createHttpMiddleware({
-          host: 'https://api.sphere.io',
-          includeOriginalRequest: true,
-          fetch,
+    it('should send an invalid client credentials', async () => {
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth.it.js]: Updated here 3a670fe

File: integration-tests/sdk/auth.it.js [hunk 6]
```diff
@@ -1,241 +1,265 @@
...
-          headers: {
-            Authorization: `Basic ${basicAuth}`,
-            'Content-Length': Buffer.byteLength(body).toString(),
-            'Content-Type': 'application/x-www-form-urlencoded',
-          },
-          body,
+      }
+    })
+
+    it('should send an invalid scope', async () => {
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth.it.js]: Updated here 3a670fe

File: packages/sdk-auth/src/auth.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+   * {
+   *   "host": "https://auth.commercetools.com",
+   *   "projectKey": "sample-project",
+   *   "credentials": {
+   *     "clientId": "sampleClient",
+   *     "clientSecret": "sampleSecret"
+   *   },
+   *   "scopes": [
+   *     "view_products:{project-key}"
+   *   ]
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: Updated here 3a670fe

File: packages/sdk-auth/src/auth.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+    const { projectKey } = config
+    this.ANONYMOUS_FLOW_URI = `/oauth/${projectKey}/anonymous/token`
+    this.PASSWORD_FLOW_URI = `/oauth/${projectKey}/customers/token`
+    this.BASE_AUTH_FLOW_URI = '/oauth/token'
+    this.INTROSPECT_URI = '/oauth/introspect'
+  }
+
+  static _getFetcher(configFetch: ?ConfigFetch): ConfigFetch {
+    return configFetch || global.fetch || isomorphicFetch
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: I was wondering here, why do we use a default fetch method when running in the browser (using a window.fetch) but when running in node we have to provide our own fetch method?

Shouldn't we simplify our SDK as much as possible and offer our customers a default configuration (scopes, fetch method)? I think that the majority of customers do not want to write their own fetch and just want to call our API.

The `isomorphicFetch` returns a fetch method based on its environment (browser/node).

File: integration-tests/sdk/auth.it.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth.it.js]: updated here 4eea819

File: packages/sdk-auth/src/auth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (daern91) [reply packages/sdk-auth/src/auth.js]: Hmm.. very good question. @tdeekens, if I remember correctly we changed this half a year ago or so. Do you remember the reasoning for it?

File: packages/sdk-auth/src/auth.js [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+  AuthMiddlewareOptions,
+  ClientAuthOptions,
+  ConfigFetch,
+  HttpErrorType,
+  AuthRequest,
+  UserAuthOptions,
+} from 'types/sdk'
+import isomorphicFetch from 'isomorphic-fetch'
+import { getErrorByCode } from '@commercetools/sdk-middleware-http'
+import * as authScopes from './scopes'
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: Updated here 19f3e13

File: integration-tests/sdk/auth-middleware.it.js [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+import { clearData, createData } from '../cli/helpers/utils'
+
+let projectKey
+if (process.env.CI === 'true') projectKey = 'auth-integration-test'
+else projectKey = process.env.npm_config_projectkey
+
+describe('Auth Middleware Flows', () => {
+  let apiConfig
+  const userEmail = `user+date_is/\\${Date.now()}@commercetooler.com`
+  const userPassword = '4^lks*aJ@ETso+/\\HdE1!x0u4q5'
```
[the remark this answers is not in the document]
Comment (junajan) [reply integration-tests/sdk/auth-middleware.it.js]: This is really for testing purposes - to test escaping of special characters in HTTP request body.. I prefixed it with `testing` - 19f3e13

File: packages/sdk-auth/src/auth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: This is still sad.. on browsers it uses the global fetch but in Node.js where there is no `fetch` method, we have to provide our own.

File: integration-tests/sdk/auth-middleware.it.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (tdeekens) [reply integration-tests/sdk/auth-middleware.it.js]: Thanks :)

File: packages/sdk-auth/src/auth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (tdeekens) [reply packages/sdk-auth/src/auth.js]: Yes kind of. On the other hand, it broke browsers serving them a Node.JS version of isomorphic fetch _and_ bloated the bundle as most browser have a native implementation already. I think for Node.JS consumers they can either pass e.g. `node-fetch` as an arg or `global.fetch = require('node-fetch')`.

[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: Please check this change - I updated it so it now works as in auth middleware - d9c9496

File: packages/sdk-auth/src/auth.js [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    this.fetcher = SdkAuth._getFetcher(config.fetch)
+
+    const { projectKey } = config
+    this.ANONYMOUS_FLOW_URI = `/oauth/${projectKey}/anonymous/token`
+    this.PASSWORD_FLOW_URI = `/oauth/${projectKey}/customers/token`
+    this.BASE_AUTH_FLOW_URI = '/oauth/token'
+    this.INTROSPECT_URI = '/oauth/introspect'
+  }
+
+  static _getFetcher(configFetch: ?ConfigFetch): ConfigFetch {
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: updated here d9c9496

File: packages/sdk-auth/src/auth.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+  // Set flowtype annotations
+  config: AuthMiddlewareOptions
+  fetcher: ConfigFetch
+  ANONYMOUS_FLOW_URI: string
+  BASE_AUTH_FLOW_URI: string
+  PASSWORD_FLOW_URI: string
+  INTROSPECT_URI: string
+
+  /**
+   * Sample configuration object:
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: Hi @adnasa .. we have it already here https://github.com/commercetools/nodejs/pull/866/files#diff-42996921e523e4a61f03925270f7dc94R38

File: packages/sdk-auth/src/auth.js [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+   * }
+   * @param config
+   */
+  constructor(config: AuthMiddlewareOptions) {
+    // validate config properties
+    SdkAuth._checkRequiredConfiguration(config)
+    this.config = config
+    this.fetcher = SdkAuth._getFetcher(config.fetch)
+
+    const { projectKey } = config
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: :-) make sense .. ca69c80

File: packages/sdk-auth/src/auth.js [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    const response = await this._performRequest(request)
+    return SdkAuth._handleResponse(request.uri, response)
+  }
+
+  static _createResponseError(
+    { message, ...rest }: Object,
+    uri: string,
+    statusCode: number
+  ): HttpErrorType {
+    let errorMessage = message || 'Unexpected non-JSON error response'
```
[the remark this answers is not in the document]
Comment (junajan) [reply packages/sdk-auth/src/auth.js]: Thanks .. I haven't seen a default param in object destructor yet .. ca69c80

**Comment by junajan** [comment]
Hi @daern91 could you help me with integration tests? Do I need to link the new module somehow or do we have to publish it first?

**Comment by daern91** [comment]
Hi @junajan, for the integration test I think you just need to add `@commercetools/sdk-auth` here: https://github.com/commercetools/nodejs/blob/4948050dabd1bcfe54ad6137795ca53c002d82f0/integration-tests/package.json

**Comment by daern91** [comment]
Hi @junajan, for the integration test I think you just need to add `@commercetools/sdk-auth` here: https://github.com/commercetools/nodejs/blob/4948050dabd1bcfe54ad6137795ca53c002d82f0/integration-tests/package.json

**Comment by daern91** [comment]
Hi @junajan, for the integration test I think you just need to add `@commercetools/sdk-auth` here: https://github.com/commercetools/nodejs/blob/4948050dabd1bcfe54ad6137795ca53c002d82f0/integration-tests/package.json

**Comment by daern91** [comment]
Hi @junajan, for the integration test I think you just need to add `@commercetools/sdk-auth` here: https://github.com/commercetools/nodejs/blob/4948050dabd1bcfe54ad6137795ca53c002d82f0/integration-tests/package.json

**Comment by daern91** [comment]
Hi @junajan, for the integration test I think you just need to add `@commercetools/sdk-auth` here: https://github.com/commercetools/nodejs/blob/4948050dabd1bcfe54ad6137795ca53c002d82f0/integration-tests/package.json

**Comment by junajan** [comment]
test

**Comment by daern91** [comment]
cc @lauraluiz 👍 😊

**Pull request merged**